### PR TITLE
feat: implement pre/post processors and hooks in async subscription chain

### PR DESCRIPTION
**Pull request by marcambier**
## Issue

https://gravitee.atlassian.net/browse/APIM-374

## Description

Here is the rendering of pre/post processors of a webhook subscription processing, in Jaeger :
image

🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.z6.web.core.windows.net/apim374-subscription-processors-and-hooks/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-fudmurpjib.chromatic.com)

### CODE REVIEW

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/jupiter/reactor/processor/DefaultPlatformProcessorChainFactory.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import io.gravitee.gateway.jupiter.reactor.processor.tracing.TraceContextProcessor;
+import io.gravitee.gateway.jupiter.reactor.processor.transaction.TransactionProcessorFactory;
+import io.gravitee.gateway.report.ReporterService;
+import io.gravitee.node.api.Node;
+import io.gravitee.plugin.alert.AlertEventProducer;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * The default {@link AbstractPlatformProcessorChainFactory} used for synchronous HTTP requests
```
Comment (guillaumelamirand) [inline gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/jupiter/reactor/processor/DefaultPlatformProcessorChainFactory.java]: This factory is also used for event native api, I would more say "used with default http dispatcher" or something like that.

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/jupiter/reactor/v4/subscription/DefaultSubscriptionDispatcher.java [hunk 2]
```diff
@@ -50,21 +63,37 @@ public class DefaultSubscriptionDispatcher extends AbstractService<SubscriptionD
 
     public static final int ON_SUBSCRIPTION_ERROR_RETRY_COUNT = 5;
     public static final int ON_SUBSCRIPTION_ERROR_RETRY_DELAY_MS = 3_000;
+
     private static final Logger LOGGER = LoggerFactory.getLogger(DefaultSubscriptionDispatcher.class);
     private static final String SUBSCRIPTION_ENTRYPOINT_FIELD = "entrypointId";
+
+    protected static final String TRACING_SPAN_NAME = "SUBSCRIPTION";
+    protected static final String ATTR_INTERNAL_TRACING_SPAN = "subscription-tracing-span";
```
Comment (guillaumelamirand) [inline gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/jupiter/reactor/v4/subscription/DefaultSubscriptionDispatcher.java:71]: Is this attribute used somewhere else? If so, could you move it to InternalContextAttribute as Constant?

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/jupiter/reactor/v4/subscription/DefaultSubscriptionDispatcher.java [hunk 3]
```diff
@@ -50,21 +63,37 @@ public class DefaultSubscriptionDispatcher extends AbstractService<SubscriptionD
...
+        SubscriptionExecutionRequestFactory subscriptionExecutionRequestFactory,
+        SubscriptionPlatformProcessorChainFactory platformProcessorChainFactory,
+        boolean tracingEnabled,
+        Vertx vertx
     ) {
         this.subscriptionAcceptorResolver = subscriptionAcceptorResolver;
         this.subscriptionExecutionRequestFactory = subscriptionExecutionRequestFactory;
+        this.platformProcessorChainFactory = platformProcessorChainFactory;
+        this.tracingEnabled = tracingEnabled;
+        this.processorChainHooks = tracingEnabled ? List.of(new TracingHook("processor-chain")) : new ArrayList<>();
```
Comment (guillaumelamirand) [inline gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/jupiter/reactor/v4/subscription/DefaultSubscriptionDispatcher.java:95]: I am not sure to understand properly; Is there a typo in the hook's name?

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactor/spring/ReactorConfiguration.java [hunk 4]
```diff
@@ -114,7 +115,7 @@ public io.gravitee.gateway.jupiter.reactor.processor.transaction.TransactionProc
     }
 
     @Bean
-    public PlatformProcessorChainFactory globalProcessorChainFactory(
+    public DefaultPlatformProcessorChainFactory globalProcessorChainFactory(
```
Comment (guillaumelamirand) [inline gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactor/spring/ReactorConfiguration.java]:
```suggestion
    public DefaultPlatformProcessorChainFactory defaultPlatformProcessorChainFactory(
```

**Review by guillaumelamirand — CHANGES_REQUESTED** [review]

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/jupiter/reactor/v4/subscription/DefaultSubscriptionDispatcher.java (hunk 2 above)
Comment (marcambier) [reply gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/jupiter/reactor/v4/subscription/DefaultSubscriptionDispatcher.java:71]: No, it's used internally only (and in junit test)

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactor/spring/ReactorConfiguration.java (hunk 4 above)
Comment (marcambier) [reply gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/reactor/spring/ReactorConfiguration.java]: :+1: done

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/jupiter/reactor/processor/DefaultPlatformProcessorChainFactory.java (hunk 1 above)
Comment (marcambier) [reply gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/jupiter/reactor/processor/DefaultPlatformProcessorChainFactory.java]: :+1: done

File: gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/jupiter/reactor/v4/subscription/DefaultSubscriptionDispatcher.java (hunk 3 above)
Comment (marcambier) [reply gravitee-apim-gateway/gravitee-apim-gateway-reactor/src/main/java/io/gravitee/gateway/jupiter/reactor/v4/subscription/DefaultSubscriptionDispatcher.java:95]: The id of the hook is `hook-tracing`, defined in TracingHook.
`processor-chain` is the key of the TracingHook, that will be used to create span names like, for example `processor-chain-before-api-handle` (as we can see in screenshot).
Same key than DefaultHttpRequestDispatcher tracingHook.
Sound relevant to you ?

**Review by guillaumelamirand — APPROVED** [review]

**Pull request merged**